File: day3/main.clj
(let [content (slurp "input.txt")
      re #"mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)"]
  (reduce
   (fn [{:keys [enabled sum cond-sum]} cmd]
     (cond
       (= (first cmd) "do()") {:enabled true :sum sum :cond-sum cond-sum}
       (= (first cmd) "don't()") {:enabled false :sum sum :cond-sum cond-sum}
       :else (let [[_ x y] cmd
                   x (Integer. x)
                   y (Integer. y)
                   product (* x y)]
               {:enabled enabled :sum (+ sum product) :cond-sum (if enabled (+ cond-sum product) cond-sum)})))
   {:enabled true :sum 0 :cond-sum 0}
   (re-seq re content)))

File: day3/src/main.rs
use regex::Regex;

fn main() {
    let re = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)").unwrap();
    let content = std::fs::read_to_string("input.txt").unwrap();
    let result = re
        .captures_iter(&content)
        .fold((true, 0, 0), |(enabled, sum, cond_sum), m| match &m[0] {
            "do()" => (true, sum, cond_sum),
            "don't()" => (false, sum, cond_sum),
            _ => {
                let a: i64 = m[1].parse().unwrap();
                let b: i64 = m[2].parse().unwrap();
                (
                    enabled,
                    sum + a * b,
                    cond_sum + if enabled { a * b } else { 0 },
                )
            }
        });
    println!("Day3.1: {}", result.1);
    println!("Day3.2: {}", result.2);
}
